src/client.rs

```diff
@@ -216,13 +206,40 @@ impl Default for Builder {
     }
 }
 
+// ===== impl Connection =====
+
+impl<T, B> Future for Connection<T, B>
+where
+    T: AsyncRead + AsyncWrite,
+    B: IntoBuf,
+{
+    type Item = ();
+    type Error = ::Error;
+
+    fn poll(&mut self) -> Poll<(), ::Error> {
+        self.inner.poll().map_err(Into::into)
+    }
+}
+
+impl<T, B> fmt::Debug for Connection<T, B>
+where
+    T: AsyncRead + AsyncWrite,
+    T: fmt::Debug,
+    B: fmt::Debug + IntoBuf,
+    B::Buf: fmt::Debug,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(&self.inner, fmt)
+    }
+}
+
 // ===== impl Handshake =====
 
 impl<T, B: IntoBuf> Future for Handshake<T, B>
 where
     T: AsyncRead + AsyncWrite,
 {
-    type Item = Client<T, B>;
+    type Item = (Connection<T, B>, Client<B>);
```

Comment:
So, I don't really care about the order between `(Connection, Client)`, but I would like to try to be consistent.

I've been using `(Self, Other)` when adding a fn in a `impl Self`. 

That said, it seems more idiomatic in Rust to have fns that return tuples to be free fns (like `mpsc`). Though, I felt that was strange when the module was the same name as one of the types being returned...